macros/timeHistoryAnalyzer.py

```diff
@@ -0,0 +1,457 @@
+    noise < threshold. During the search, at most maxSkip scans with mask not
+    set can be skipped. Only ranges with more than minBadScans are kept.
+
+    Args:
+        data: The TimeSeriesData object to pull data from
+        vfat: The VFAT to return ranges for
+        channel: The channel to return ranges for
+        threshold: The minimum value of noise for a scan to be "good"
+        maxSkip: The maximum number of "good" scans between two "bad" scans
+        minBadScans: The minimum number of bad scans
+
+    Returns:
+        A list of MaskedRange objects
+    """
+    ranges = _findRangesMeta(data,
+                             vfat,
+                             channel,
+                             data.noise[vfat,channel] < threshold,
+                             maxSkip)
+
+    return list(filter(lambda r: np.count_nonzero(r.noise() < threshold) >= minBadScans,
+                       ranges))
+
+class TimeSeriesData(object):
+    """Holds information about time variation of scan results.
+
+    Each property is stored as a 3D Numpy array with indexes
+    [vfat][strip][time]. The time index -> scan date mapping is exposed in the
+    date attribute.
+
+    Attributes:
+        dates: Array of strings containing the scan dates.
+        mask: Data for the "mask" property,
+        maskReason: Data for the "maskReason" property,
+    """
+
+    def __init__(self, inputDir):
+        """Creates a TimeSeriesData object by reading the files located in the
+        inputDir directory.
+
+        The input directory must contain the following files:
+
+        * gemPlotterOutput_mask_vs_scandate.root
+        * gemPlotterOutput_maskReason_vs_scandate.root
+
+        They are created by plotTimeSeries.py.
+
+        Args:
+            inputDir: The path to the input directory
+        """
+        import ROOT as r
+        from root_numpy import hist2array
+
+        file_mask = r.TFile('%s/gemPlotterOutput_mask_vs_scandate.root' % inputDir, 'READ')
+        file_maskReason = r.TFile('%s/gemPlotterOutput_maskReason_vs_scandate.root' % inputDir, 'READ')
+        file_noise = r.TFile('%s/gemPlotterOutput_noise_vs_scandate.root' % inputDir, 'READ')
+
+        self.mask = [] # [vfat][time][strip]; warning: reordered after loading
+        self.maskReason = [] # [vfat][time][strip]; warning: reordered after loading
+        self.noise = [] # [vfat][time][strip]; warning: reordered after loading
+
+        for vfat in range(0,24):
+            dirname = 'VFAT%d' % vfat
+            dir_mask = file_mask.Get(dirname)
+            dir_maskReason = file_maskReason.Get(dirname)
+            dir_noise = file_noise.Get(dirname)
+
+            hist_mask = dir_mask.Get("h_ROBstr_vs_scandate_Obsmask_VFAT%d" % vfat)
+            hist_maskReason = dir_maskReason.Get("h_ROBstr_vs_scandate_ObsmaskReason_VFAT%d" % vfat)
+            hist_noise = dir_noise.Get("h_ROBstr_vs_scandate_Obsnoise_VFAT%d" % vfat)
```

Comment:
why bother creating a `TDirectory` object here?  You can get these `TObjects` directly from the `TFile` object without needing to have an additional set of objects. e.g.:

```
            hist_noise = file_noise.Get("VFAT%i/h_ROBstr_vs_scandate_Obsnoise_VFAT%i" %(vfat, vfat))
```

Additionally change `%d` to `%I`; `vfat` is an integer not a floating point number; this will ensure no funny business (e.g. `1.0` vs. `1` is inserted).

Reply:
[`%d` = `%i` = "Signed integer decimal"](https://docs.python.org/2/library/stdtypes.html#string-formatting-operations)